bin/tx-prover/src/prover.rs

```diff
@@ -11,39 +14,48 @@ use crate::{generated::api_client::ApiClient, RemoteTransactionProverError};
 
 /// A [RemoteTransactionProver] is a transaction prover that sends witness data to a remote
 /// gRPC server and receives a proven transaction.
+///
+/// When compiled for the `wasm32-unknown-unknown` target, it uses the `tonic_web_wasm_client`
+/// transport. Otherwise, it uses the built-in `tonic::transport` for native platforms.
+/// The transport layer connection is established lazily when the first transaction is proven.
 #[derive(Clone)]
 pub struct RemoteTransactionProver {
     #[cfg(target_arch = "wasm32")]
-    client: RefCell<ApiClient<tonic_web_wasm_client::Client>>,
+    client: RefCell<Option<ApiClient<tonic_web_wasm_client::Client>>>,
 
     #[cfg(not(target_arch = "wasm32"))]
-    client: RefCell<ApiClient<tonic::transport::Channel>>,
+    client: RefCell<Option<ApiClient<tonic::transport::Channel>>>,
+
+    endpoint: String,
 }
 
 impl RemoteTransactionProver {
-    /// Creates a new [RemoteTransactionProver] with the specified gRPC server endpoint.
-    /// This instantiates a tonic client that attempts connecting with the server.
-    ///
-    /// When compiled for the `wasm32-unknown-unknown` target, it uses the `tonic_web_wasm_client`
-    /// transport. Otherwise, it uses the built-in `tonic::transport` for native platforms.
-    ///
-    /// # Errors
-    ///
-    /// This function will return an error if the endpoint is invalid or if the gRPC
-    /// connection to the server cannot be established.
-    pub async fn new(endpoint: &str) -> Result<Self, RemoteTransactionProverError> {
+    /// Creates a new [RemoteTransactionProver] with the specified gRPC server endpoint. The
+    /// endpoint should be in the format `{protocol}://{hostname}:{port}`.
+    pub fn new(endpoint: &str) -> Self {
+        RemoteTransactionProver {
+            endpoint: endpoint.to_string(),
+            client: RefCell::new(None),
+        }
+    }
+
+    async fn connect(&self) -> Result<(), RemoteTransactionProverError> {
```

Comment:
Let's add comments to this function.

Also, I would probably move the check for whether we are connected or not here. Basically, this method would first check if we are already connected, and if not, connect.